### DISCUSSION TITLE
Which .so files are safe to remove from a packaged PyQt5 app?

### DISCUSSION THREAD

**Discussion by ErikBjare** [pyinstaller/pyinstaller#6645] [Help] [answered]
Hi there, I'm building [ActivityWatch](https://github.com/ActivityWatch/activitywatch) and we've been using PyInstaller for years now.

One issue we've had from time to time are crashes caused by out-of-date libraries in our build environment, that when packaged, results in errors on users systems.

We've been able to work around these issues by simply removing offending .so files from the package, but they seem to [pop up one after one](https://github.com/ActivityWatch/activitywatch/blob/50967175e4d27d8d9dcbdef4dc4b5c3825005188/Makefile#L164-L171)...

So, currently we remove:
 - libdrm.so.2
 - libharfbuzz.so.0
 - libfontconfig.so.1
 - libfreetype.so.6

And we will probably be removing `libgio-2.0.so.0` too: https://github.com/ActivityWatch/activitywatch/issues/710

Now I'm wondering, I can't be alone here, so are there any others I can preemptively remove, without causing trouble?

**Comment by rokm** [comment, marked as answer]
There's no simple answer here, as the situation depends on your build distribution and your target distribution, and the (in)compatibility between their shared libraries, both in terms of ABI (i.e., library versions), and the distribution-specific (patched) behavior.

**Reply by ErikBjare** [reply]
Right, but is there not a list/set of shared libraries that can (usually) be safely left out? I get libfreetype, libfontconfig, libharfbuzz, since all distros should have them. 

But what about libgio for example? Does deleting it seem reasonable? I'm more interested in a process to follow rather than individual answers.

**Reply by bwoodsend** [reply]
Sticking:
```python
from psutil import Process
libraries_used = [i.path for i in Process().memory_maps()]
```
at the end of your code then comparing that to what's in your bundle would be the process but there will be unloaded libraries that you must ship anyway because they are only loaded on wayland or X.

**Reply by rokm** [reply]
> Right, but is there not a list/set of shared libraries that can (usually) be safely left out?

If you come across a universal list like that, do let me know...

> I get libfreetype, libfontconfig, libharfbuzz, since all distros should have them.

What you mean is that you assume that all distros provide *compatible versions* of these libraries. 

Which may or may not be the case at the moment (in the current "vintage" of the distributions), and might increasingly be less the case as time goes on. I.e., if you build on old enough distribution, the newer versions of the removed libraries provided by target system may satisfy the older-version requirements of libraries collected from the build system. On the other hand, the newer system-provided version of the library on target system may break the ABI in its newer version, and things will stop working there...

These changes in ABI usually do not pose a problem for linux distributions themselves, because all their software is (re)built against that particular version of the library, but it does make life difficult for binary-only software. There's a reason why games on Steam nominally target (a specific version of) Ubuntu, and why common-runtime approaches, such as Flatpak, have been developed.

> But what about libgio for example? Does deleting it seem reasonable?

Do you plan to run the frozen application on a distribution that provides older version of GLib/Gio than the one on your build distribution?

As a side note, if you consider removing `libgio`, you should probably also consider removing all other collected GLib libraries. I imagine just having those at mixed versions might cause some problems...

> I'm more interested in a process to follow rather than individual answers.

I suppose you could check the linux PyQt5 wheel(s) and what libraries they do provide in addition to Qt5 ones. Because everything else that's not there is assumed to be on the host system already. (That's assuming you aren't using other packages that may pull in GLib, X, various image/multimedia libraries, etc.).

But ultimately, you'll have to decide both on what distributions you wish to support/target, and what you expect to find there. And then remove the libraries you think can be removed, and test if the things still work on those distributions you have selected.

**Reply by ErikBjare** [reply]
@bwoodsend Neat, that snippet will come in handy.

@rokm Excellent answer, thank you very much.
